### ISSUE TITLE
How to inherit?

### ISSUE THREAD

**Issue by goswinr** [glutinum-org/Glutinum.Converter#16]
TS can inherit interfaces, F# not.
This can be bypassed by creating a separate module that has classes of the same name:

Generated code from [Types.Web ](https://www.npmjs.com/package/@types/web):

```fsharp
open Fable.Core

/// All classes in this module are empty.
/// They are only used to allow inheriting from types defined as interface.
/// Use this module when you need to inherit from type that is only defined as an interface.
[<RequireQualifiedAccess>] // don't open this. it would shadow the interfaces
module Inheritable =
    

    /// This is NOT a Browser.Types.HTMLElement !
    /// It is just an empty class with the same name.
    /// It used only for inheriting the Browser.Types.HTMLElement interfaces in F#.
    /// Use like: ' type [<AttachMembers>] MyClass() = inherit Inheritable.HTMLElement()'
    /// Then inside the class you can get the current instance via 'base.this' 
    [<Global>]    
    type HTMLElement() = 

        [<Global>]
        member _.this : Browser.Types.HTMLElement = 
            JsInterop.emitJsExpr () "this" 
```
In user code a [Web Component](https://developer.mozilla.org/en-US/docs/Web/API/Web_components) can be made like this: 

```fsharp
open Fable.Core

[<AttachMembers>] 
type MyWebComponent() =
    inherit Inheritable.HTMLElement()     
    let myElem = base.this // typed as Browser.Types.HTMLElement not Inheritable.HTMLElement
```

this gnerates: 
```javascript
  export class MyWebComponent extends HTMLElement {
      constructor() {
          super();
          const myElem = this;
      }
  }
```

see [REPL](https://fable.io/repl/#?code=PYBwpgdgBAYghgIwDZgHQGFgCcwCh8D0RUAgkklAMZJwDOtYtUAltAC4AWzTAtsACYBXFFDg4oYHiDYBPVLiIEoAFQ5gZo8cAhINghvyhtgo8sADuLCGqzM2rAOZQAZlmA8jM8E35hnrMEM6KzYwLGc4SjQFYgBVBiMuXgFhMChzNWgZYEEoCDBAoxNWGzsXNw9ZcES4NhYmbV0oX398oKY4aFZQ8MjogG0AHgAlMABHQWYcAEVBOCRmf0CSSij6AD4AXShifm0AcjrQSETuVBY68xykQ1oOOD3LTjTusIi13D4hEQBJazC7IgRABeXBQcFQfAQnbEVTcepQAByAHllKIoAAhNzmBhYVDKLyMVAACWUAFkADIAURQPEgdQAhGCIYooD86vCAFb6OqdCRSWRUGj0dJ2DiJNK0OB0vLS6LQ1nsqD6QqNDTObBWUr2CBOZ6Y7G4-GE2gk8nU2n0kJvPpMViwADE8gVcQSCwA1mAAFxQfaeapDEhsNiRDhkyQIMK0LZQMkydDC2gACgAlFBgVqAXU-tqgWhSZSaZJ6an9szwazVCdWLRmL4JUK6ExsrlKHyHGA6vrKIIsDgIByILQQxAolAAG7MOC+hB0NCcbh+8v+tJDADiSGAs6QMYLFuLA9T6chy+h683262p4hdJ4kawUAA+qgF0wfViLEaCd4zYXLQPj2vaFwQAKVoP4elAVBJDsMCqQADxAB8jwAIlfFCT3wQNg1DcM7yjGNcCqNI4wAdTABBMCkbQSzTUFoRKLM2X+WwQ2QfNzSLOlDzTaFlxQOoeBkLjj1nBgXySGEV3aA1PzCY0fz3LirQgYBsxYwF2N-fduLYXAgA&html=DwCwLgtgNgfAsAKAAQqaApgQwCb2ag4CdMTJcMABwFp0BHAVwEsA3AXgCIBhAewDsw6AdQAqAT0roOSAMb9BAzoIAeYAPThoAbhkhMAJwDOJNgzAAzagA4OeQhqy5EhAEY9sYu6mBq3HvD6asEA&css=Q)

**Comment by MangelMaxime** [comment]
This is an interesting idea. 👍 

Could you please provide the TypeScript snippet that correspond to what we are discussing? This is because it helps visualise what the TypeScript code is and how each element should transformed into F#.

Does this means that for each interface we need to generate a corresponding `Inheritable` type? If yes, is there situation where we don't want to?

If I understand correctly, this means that `Browser.Types.HTMLElement` can still inherit from mutile interface because it is an interface.

And the `Inheritable` types are there just for specialisation for example when using Web Component?

When using Web Component (not limited to), is there cases where we want to inherit from several element or is it always only one?

**Comment by goswinr** [comment]
Yes! You are right. This applies to any interface. On your last question, I think JS class can only inherit from one class. (I will add  some more snippets later..)

**Comment by MangelMaxime** [comment]
Something I just thought, is that because of the line `JsInterop.emitJsExpr () "this"` then generated code is not a pure binding (it contains code).

Luckily using the following:

```fs
[<Global>]
member _.this : Browser.Types.HTMLElement = 
	nativeOnly
```

generates the same code and means that we have a pure binding.

Having pure binding is important because otherwise, you need to include the F# files in the `fable` folder, which means that Fable will use the sources files instead of the DLL for the compilation. Using the DLL is much faster because the F# compiler has less to do.
